### PR TITLE
Use loopback interface redirection

### PR DESCRIPTION
**Pull request by alexstrat**
- [x] add option to customize port and sucessRedirectURL
- [x] add option `focusAfterSuccess`

### CODE REVIEW

**Comment by alexstrat** [comment]
Published as `2.0.0-rc`

File: src/index.ts [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
 import { Credentials } from 'google-auth-library/build/src/auth/credentials';
 import { google } from 'googleapis';
 import { stringify } from 'querystring';
 import * as url from 'url';
+import LoopbackRedirectServer from './LoopbackRedirectServer';
 
 const BW: typeof BrowserWindow = process.type === 'renderer' ? remote.BrowserWindow : BrowserWindow;
 
+// can't be randomozed
+const LOOPBACK_INTERFACE_REDIRECTION_PORT = 42813;
```
Comment (magne4000) [inline src/index.ts]: I know that we cannot randomize it yet, but shouldn't this hard coded value be one layer up (not inside this project)?

File: src/index.ts [hunk 2]
```diff
@@ -1,14 +1,18 @@
...
 import { OAuth2Client } from 'google-auth-library';
 import { Credentials } from 'google-auth-library/build/src/auth/credentials';
 import { google } from 'googleapis';
 import { stringify } from 'querystring';
 import * as url from 'url';
+import LoopbackRedirectServer from './LoopbackRedirectServer';
 
 const BW: typeof BrowserWindow = process.type === 'renderer' ? remote.BrowserWindow : BrowserWindow;
 
+// can't be randomozed
```
Comment (magne4000) [inline src/index.ts]: _randomozed_

File: src/index.ts [hunk 3]
```diff
@@ -90,51 +103,42 @@ export default class ElectronGoogleOAuth2 extends EventEmitter {
...
-    });
-  }
+    // waitForRedirection will close the server
+    this.server = null;
 
+    const parsed = url.parse(reachedCallbackURL, true);
+    if (parsed.query.error) {
+      throw new Error(parsed.query.error_description as string);
+    } else if (!parsed.query.code) {
+      throw new Error('Unknown' as string);
```
Comment (magne4000) [inline src/index.ts]: `as string` is probably not necessary here

File: src/index.ts [hunk 4]
```diff
@@ -90,51 +103,42 @@ export default class ElectronGoogleOAuth2 extends EventEmitter {
...
+    }
+
+    // refocus on the window
+    BW.getAllWindows().filter(w => w.isVisible()).forEach(w => w.show());
```
Comment (magne4000) [inline src/index.ts]: Should we put this logic behind an option, like `focusAfterSuccess: boolean = true`?

**Review by magne4000 — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 4 above)
Comment (magne4000) [reply src/index.ts]: Or trigger a custom event that we can listen on

File: src/index.ts (hunk 1 above)
Comment (alexstrat) [reply src/index.ts]: Yes, indeed, I'll add some options and leave a default value.

File: src/index.ts (hunk 4 above)
Comment (alexstrat) [reply src/index.ts]: ok for the `focusAfterSuccess` option

File: src/index.ts (hunk 3 above)
Comment (alexstrat) [reply src/index.ts]: lol yes

**Comment by alexstrat** [comment]
Published as `2.0.0-rc.2`

**Pull request merged**